x/market/keeper/deal.go

```diff
@@ -0,0 +1,120 @@
+package keeper
+
+import (
+	"fmt"
+	sdk "github.com/cosmos/cosmos-sdk/types"
+	sdkerrors "github.com/cosmos/cosmos-sdk/types/errors"
+	authtypes "github.com/cosmos/cosmos-sdk/x/auth/types"
+	gogotypes "github.com/gogo/protobuf/types"
+	"github.com/medibloc/panacea-core/v2/x/market/types"
+)
+
+func (k Keeper) CreateNewDeal(ctx sdk.Context, owner sdk.AccAddress, deal types.Deal) (uint64, error) {
+	dealId := k.GetNextDealNumberAndIncrement(ctx)
+
+	newDeal, err := k.newDeal(dealId, deal)
+	if err != nil {
+		return 0, err
+	}
+
+	var coins sdk.Coins
+	coins = append(coins, *deal.GetBudget())
+
+	acc := k.accountKeeper.GetAccount(ctx, sdk.AccAddress(newDeal.GetDealAddress()))
+	if acc != nil {
+		return 0, sdkerrors.Wrapf(types.ErrDealAlreadyExist, "deal %d already exist", dealId)
+	}
+
+	k.SetDeal(ctx, newDeal)
+
+	dealAddress, err := types.AccDealAddressFromBech32(newDeal.GetDealAddress())
+	if err != nil {
+		return 0, err
+	}
+
+	acc = k.accountKeeper.NewAccount(ctx, authtypes.NewModuleAccount(
+		authtypes.NewBaseAccountWithAddress(
+			dealAddress,
+		),
+		newDeal.DealAddress),
+	)
+	k.accountKeeper.SetAccount(ctx, acc)
+
+	err = k.bankKeeper.SendCoins(ctx, owner, dealAddress, coins)
+	if err != nil {
+		return 0, sdkerrors.Wrapf(types.ErrNotEnoughBalance, "The owner's balance is not enough to make deal")
+	}
+	return newDeal.GetDealId(), nil
+}
+
+func (k Keeper) newDeal(dealId uint64, deal types.Deal) (types.Deal, error) {
+
+	dealAddress := types.NewDealAddress(dealId)
+
+	newDeal := &types.Deal{
+		DealId:               dealId,
+		DealAddress:          dealAddress.String(),
+		DataSchema:           deal.GetDataSchema(),
+		Budget:               deal.GetBudget(),
+		TrustedDataValidator: deal.GetTrustedDataValidator(),
+		WantDataCount:        deal.GetWantDataCount(),
+		CompleteDataCount:    0,
+		Owner:                deal.GetOwner(),
+		Status:               "ACTIVE",
```

Comment:
I think it would be better set `const`

```go
const (
  ACTIVE      = "ACTIVE"
  DEACTIVE    = "DEACTIVE"
  COMPLETED    = "COMPLETED"
)
```

Reply:
Yes, I agree about this. I'll status field to ```const```.